openedx_learning/apps/authoring/publishing/models/draft_log.py

```diff
@@ -0,0 +1,307 @@
+
+    Our first use case for this is that changes involving child components are
+    thought to affect parent Units, even if the parent's version doesn't change.
+
+    Side-effects are recorded in a collapsed form that only captures one level.
+    So if Components C1 and C2 are both changed and they are part of Unit U1,
+    which is in turn a part of Subsection SS1, then the DraftSideEffect entries
+    are::
+
+      (C1, U1)
+      (C2, U1)
+      (U1, SS1)
+
+    We do not keep entries for (C1, SS1) or (C2, SS1). This is to make the model
+    simpler, so we don't have to differentiate between direct side-effects and
+    transitive side-effects in the model.
+
+    We will record side-effects on a parent container whenever a child changes,
+    even if the parent container is also changing in the same DraftChangeLog.
+    The child change is still affecting the parent container, whether the
+    container happens to be changing for other reasons as well. Whether a parent
+    -child relationship exists or not depends on the draft state of the
+    container at the *end* of a bulk_draft_changes_for context. To give concrete
+    examples:
+
+    Setup: A Unit version U1.v1 has defined C1 to be a child. The current draft
+    version of C1 is C1.v1.
+
+    Scenario 1: In the a bulk_draft_changes_for context, we edit C1 so that the
+    draft version of C1 is now C1.v2. Result:
+    - a DraftChangeLogRecord is created for C1.v1 -> C1.v2
+    - a DraftChangeLogRecord is created for U1.v1 -> U1.v1
+    - a DraftSideEffect is created with cause (C1.v1 -> C1.v2) and effect
+      (U1.v1 -> U1.v1). The Unit draft version has not been incremented because
+      the metadata a Unit defines for itself hasn't been altered, but the Unit
+      has *changed* in some way because of the side effect of its child being
+      edited.
+
+    Scenario 2: In a bulk_draft_changes_for context, we edit C1 so that the
+    draft version of C1 is now C1.v2. In the same context, we edit U1's metadata
+    so that the draft version of U1 is now U1.v2. U1.v2 still lists C1 as a
+    child entity. Result:
+    - a DraftChangeLogRecord is created for C1.v1 -> C1.v2
+    - a DraftChangeLogRecord is created for U1.v1 -> U1.v2
+    - a DraftSideEffect is created with cause (C1.v1 -> C1.v2) and effect
+      (U1.v1 -> U1.v2)
+
+    Scenario 3: In a bulk_draft_changes_for context, we edit C1 so that the
+    draft version of C1 is now C1.v2. In the same context, we edit U1's list of
+    children so that C1 is no longer a child of U1.v2. Result:
+    - a DraftChangeLogRecord is created for C1.v1 -> C1.v2
+    - a DraftChangeLogRecord is created for U1.v1 -> U1.v2
+    - no SideEffect is created, since changing C1 does not have an impact on the
+      current draft of U1 (U1.v2). A DraftChangeLog is considered a single
+      atomic operation, so there was never a point at which C1.v1 -> C1.v2
+      affected the draft state of U1.
+    """
```

Comment:
Note to self: Add a scenario where a change is effectively "undone", so no side-effect needs to be created, e.g. (C1.v1 -> C1.v2) followed by (C1.v1 -> C1.v1) in the same context means that nothing has changed from before the context to after the context. As such, it had no actual side effect on U1, and we should just prune things like this from the change log without creating a SideEffect entry.